proposals/user-defined-compound-assignment.md

```diff
@@ -0,0 +1,568 @@
+# User Defined Compound Assignment Operators
+
+## Summary
+[summary]: #summary
+
+Allow user types to customize behavior of compound assignment operators in a way that the target of the assignment is
+modified in-place.
+
+## Motivation
+[motivation]: #motivation
+
+C# provides support for the developer overloading operator implementations for user-defined type.
+Additionally, it provides support for "compound assignment operators" which allow the user to write
+code similarly to `x += y` rather than `x = x + y`. However, the language does not currently allow
+for the developer to overload these compound assignment operators and while the default behavior
+does the right thing, especially as it pertains to immutable value types, it is not always
+"optimal".
+
+Given the following example
+``` C#
+class C1
+{
+    static void Main()
+    {
+        var c1 = new C1();
+        c1 += 1;
+        System.Console.Write(c1);
+    }
+    
+    public static C1 operator+(C1 x, int y) => new C1();
+}
+```
+
+with the current language rules, compound assignment operator `c1 += 1` invokes user defined `+` operator
+and then assigns its return value to the local variable `c1`. Note that operator implementation must allocate
+and return a new instance of `C1`, while, from the consumer's perspective, an in-place change to the original
+instance of `C1` instead would work as good (it is not used after the assignment), with an additional benefit of
+avoiding an extra allocation. 
+
+When a program utilizes a compound assignment operation, the most common effect is that the original value is
+"lost" and is no longer available to the program. With types which have large data (such as BigInteger, Tensors, etc.)
+the cost of producing a net new destination, iterating, and copying the memory tends to be fairly expensive.
+An in-place mutation would allow skipping this expense in many cases, which can provide significant improvements
+to such scenarios.
+
+Therefore, it may be beneficial for C# to allow user types to
+customize behavior of compound assignment operators and optimize scenarios that would otherwise need to allocate
+and copy.
```

Comment:
It might be beneficial to give some example code of a custom compound/increment operator here before diving to the detailed design.